### ISSUE TITLE
Radius of "disk" kernel in WcsNDMap.smooth

### ISSUE THREAD

**Issue by lmohrmann** [gammapy/gammapy#1803]
An `astropy.convolution.Tophat2DKernel` is used when `kernel="disk"` in `WcsNDMap.smooth`. According to the astropy docs ([here](http://docs.astropy.org/en/stable/api/astropy.convolution.Tophat2DKernel.html#astropy.convolution.Tophat2DKernel)), the first argument of `Tophat2DKernel.__init__` should be the radius of the kernel. In `WcsNDMap.smooth`, however, `width = 2 * radius + 1` is passed for this argument. This looks like a bug to me, the radius should be passed directly here. Can someone confirm?

P.S. I'm also a bit puzzled by the `width = radius / 2.` that is used in case of the `gaussian_filter`. Why divide by 2?

**Comment by lmohrmann** [comment]
I'm pretty sure this is a bug. If I smooth a single pixel with radius=3, the result has entries up to seven pixels away (as expected, if 2*3+1=7 is passed as radius...).

The definition of the radius in case of the gaussian_filter seems to follow the ds9 convention, although I find it fairly un-intuitive that a gaussian with half the width of the specified radius is used.

**Comment by adonath** [comment]
@lmohrmann, @cdeil  Yes, the idea that we just reproduce the ds9 smoothing behaviour as documented in the section "Smoothing" in the (ds9 docs)[http://ds9.si.edu/doc/ref/how.html#Smoothing]. But I totally agree I find the convention that ds9 uses rather unintuitive. I'd be open to change the definition of the smoothing radius for `WcsNDMap.smooth` to something more intuitive. Maybe we should change to the definitions used in the `Gaussian2DKernel`, `Tophat2DKernel` and `Box2DKernel` classes in Astropy. And maybe change the keyword from `radius` to `width`, because a box shaped filter does not have a radius...

**Comment by cdeil** [comment]
Astropy should be stable, changing this there leads to chaos. In Gammapy, OK to put something else if you prefer it. For me, I think both just the most obvious with parameter would be OK (sigma for gauss, radius for disk), or some value so that the RMS width is roughly the same also. Note that this is just a little convenience method, we shouldn't spend too much time on this. The most important aspect is to document exactly what it does in the docstring; referencing the DS9 docs is OK / good, but still the description in the Gammapy docstring should be complete without the DS9 page.

**Comment by adonath** [comment]
@cdeil Just to clarify: I did not propose to change anything in Astropy, but to use the same definitions of the smoothing width, such that the following is equivalent:

```
Gaussian2DKernel(width) <=> Map.smooth("gaussian", width)  
Tophat2DKernel(width) <=> Map.smooth("disk", width)  
Box2DKernel(width) <=> Map.smooth("box", width)

**Comment by cdeil** [comment]
Ah, I misunderstood. That sounds good to me. @lmohrmann and @AtreyeeS  - thoughts?

**Comment by AtreyeeS** [comment]
I am OK with any definition as long as it is consistent. I think this suggestion is quite intuitive!

**Comment by lmohrmann** [comment]
Sounds good to me. Currently, we use scipy methods for the gaussian and box case. Should we switch to the Astropy kernels also for these cases? In any case, the radius passed to the Tophat2DKernel needs to be fixed.

**Comment by adonath** [comment]
@lmohrmann The reason why I used the Scipy methods in the initial implementation was performance only. The `gaussian_filter` and `uniform_filter` use a clever implementation based on the separability of the kernels (you first smooth along the x and then along the y axis). This way you typically get speed-ups in the order of ~100 compared to the generic convolution, which I think might be worth keeping. As compromise we could just use Astropy kernels and `Map.convolve(use_fft=True)`. 

@AtreyeeS Do you have time to work on this? You could extend my [image convolve toy benchmark notebook](https://github.com/adonath/notebooks-public/blob/master/convolution_functions_comparison.ipynb) with a section on image smoothing. Just very short, maybe spending one hour to compare the `gaussian_filter` and `uniform_filter` against `Map.convolve(kernel, use_fft=True)`  using Astropy kernels. Then we can make a decision and implement it.

**Comment by lmohrmann** [comment]
> The reason why I used the Scipy methods in the initial implementation was performance only.
Ah, I see. Then I'd suggest to keep this (rather than using fft by default). The main issue really is the wrong use of the radius for the Tophat2DKernel. If we're on it, I'd be in favour of changing the parameter definitions for all kernels as @adonath suggested.

**Comment by AtreyeeS** [comment]
Yeah I will do this today. I thought there was already an issue where it was decided scipy fftconvolve was the most convenient.

**Comment by adonath** [comment]
@AtreyeeS Yes, for the generic `Map.convolve()`we have decided that Scipy's `fftconvolve()` is currently our best choice. But the `gaussian_filter` and `uniform_filter` implement special cases of the convolution with a gauss and a box kernel. In this case you can save computing time by exploiting a property called *separability* of the kernels (see [here](https://en.wikipedia.org/wiki/Separable_filter) for an explanation). My guess is that these optimized implementations achieve better performance in most cases, maybe except for very large kernels, where `fftconvolve()` might win. So I would suggest to just change the the definition of the smoothing width but keep using the  `gaussian_filter` and `uniform_filter` methods.

**Comment by AtreyeeS** [comment]
@adonath : This separability of kernels is an interesting thing! Thanks for pointing it out.
You are absolutely correct about the runtimes. For details, do see: https://github.com/AtreyeeS/test_notebooks/blob/master/convolution_functions_comparison.ipynb

To get the with between astropy kernals, the Gaussian width still needs to be divided by 2, `Gaussian2DKernel(width) <=> Map.smooth("gaussian", width/2)`

**Comment by adonath** [comment]
Addressed in #1807
